### ISSUE TITLE
Generic TBB Thread issues in the framework and modules.

### ISSUE THREAD

**Issue by permcody** [idaholab/moose#6436]
### Description of the enhancement or error report

We are still receiving several reports of several threaded tests failing on our test systems. These need to be fix. We are engineering a way to automatically generate and retrieve "core" files from our test systems. Reference this ticket as we fix these issues.

### Rationale for the enhancement or information for reproducing the error

Many thread issues are not easily repeatable. We are trying to generate "core" files in lieu of having a good test case in which to reliably produce errors

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

Threaded runs through the framework and applications.

**Comment by permcody** [comment]
Here is one of our first stack traces produced by a TBB run on our test systems:

```
#0  0x00007ff25aa24bbc in libMesh::PetscVector<double>::get(std::vector<unsigned int, std::allocator<unsigned int> > const&, double*) const ()
   from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#1  0x00007ff25ae0aa7e in MooseVariable::computeNodalValues() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#2  0x00007ff25ad0afc0 in SystemBase::reinitNodeFace(libMesh::Node const*, short, unsigned int) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#3  0x00007ff25ad631ef in ComputeNodalAuxBcsThread::onNode(__gnu_cxx::__normal_iterator<BndNode const* const*, std::vector<BndNode const*, std::allocator<BndNode const*> > >&) ()
   from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#4  0x00007ff25ad3e242 in ThreadedNodeLoop<libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>, __gnu_cxx::__normal_iterator<BndNode const* const*, std::vector<BndNode const*, std::allocator<BndNode const*> > > >::operator()(libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*> const&) ()
   from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#5  0x00007ff25ad615c6 in void tbb::interface6::internal::partition_type_base<tbb::interface6::internal::auto_partition_type>::execute<tbb::interface6::internal::start_reduce<libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>, ComputeNodalAuxBcsThread, tbb::auto_partitioner const>, libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*> >(tbb::interface6::internal::start_reduce<libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>, ComputeNodalAuxBcsThread, tbb::auto_partitioner const>&, libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>&) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#6  0x00007ff25ad617d2 in tbb::interface6::internal::start_reduce<libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>, ComputeNodalAuxBcsThread, tbb::auto_partitioner const>::execute() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#7  0x00007ff25bb7415a in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all (this=0x7ff24fdc3000, parent=..., child=0x8)
    at ../../src/tbb/custom_scheduler.h:438
#8  0x00007ff25bb72780 in tbb::internal::generic_scheduler::local_spawn_root_and_wait (this=0x0, first=..., next=@0x8: <error reading variable>) at ../../src/tbb/scheduler.cpp:626
#9  0x00007ff25bb726ab in tbb::internal::generic_scheduler::spawn_root_and_wait (this=0x0, first=..., next=@0x8: <error reading variable>) at ../../src/tbb/scheduler.cpp:634
#10 0x00007ff25ad38f21 in AuxiliarySystem::computeNodalVars(ExecFlagType) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#11 0x00007ff25ad3b36a in AuxiliarySystem::compute(ExecFlagType) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#12 0x00007ff25ae42775 in FEProblem::computeResidualType(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, Moose::KernelType) ()
   from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#13 0x00007ff25ae23440 in FEProblem::computeResidual(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&) ()
   from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#14 0x00007ff25ad805bc in NonlinearSystem::solve() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#15 0x00007ff25ae2b285 in FEProblem::solve() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#16 0x00007ff25ab3c12e in TimeStepper::step() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#17 0x00007ff25af5c680 in Transient::solveStep(double) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#18 0x00007ff25af5c154 in Transient::takeStep(double) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#19 0x00007ff25af5bc31 in Transient::execute() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#20 0x00007ff25acd34ed in MooseApp::executeExecutioner() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#21 0x00007ff25acd87de in MooseApp::run() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#22 0x0000000000403b92 in main ()
```

**Comment by friedmud** [comment]
This particular one is happening because `PetscVector::_get_array()` is _NOT_ threadsafe: https://github.com/libMesh/libmesh/blob/master/src/numerics/petsc_vector.C#L1318

Locks need to be added in there...

@roystgnr Do you have any input on that?

**Comment by jwpeterson** [comment]
This correctly uses `VecGetArrayRead()` now, are you saying that's still not thread-safe?

**Comment by friedmud** [comment]
Setting the `_values` pointer is a race condition.  One thread might be setting the pointer while another thread is reading from it.

**Comment by friedmud** [comment]
It needs a doubly-checked lock here: https://github.com/libMesh/libmesh/blob/master/src/numerics/petsc_vector.C#L1323

**Comment by friedmud** [comment]
Also: two threads could be simultaneously _setting_ the `_values` pointer... could just lead to corrupting it.

**Comment by friedmud** [comment]
Just look at `VecGetArrayRead()`: http://www.mcs.anl.gov/petsc/petsc-current/src/vec/vec/interface/rvector.c.html#VecGetArrayRead

There is no locking in there... it just directly sets `a` (or passes it down to `VecGetArray` which sets it).

**Comment by jwpeterson** [comment]
Yeah, I'm remembering now that `VecGetArrayRead()` was just to allow this call to work when the solution vector was locked by the SNES, it has nothing to do with thread safety...

I've never implemented a "doubly-checked lock", is there an example somewhere?

**Comment by friedmud** [comment]
I can do it real quick.  Give me a minute.

**Comment by permcody** [comment]
Hopefully this fixes a lot of the thread issues we've been seeing for
*years*. That code has been that way for a long time.

**Comment by friedmud** [comment]
I have a fix implemented - testing now... but the the true test won't happen until it gets merged into MOOSE...

**Comment by brianmoose** [comment]
Here is another from https://www.moosebuild.org/job/11852/

```
#0  0x00007fca56019084 in libMesh::PetscVector<double>::operator()(unsigned int) const () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#1  0x00007fca562f2f16 in MooseVariable::computeElemValues() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#2  0x00007fca562c0774 in AuxiliarySystem::reinitElem(libMesh::Elem const*, unsigned int) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#3  0x00007fca563219d3 in FEProblem::reinitElem(libMesh::Elem const*, unsigned int) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#4  0x00007fca562bdfc1 in ComputeElemAuxVarsThread::onElement(libMesh::Elem const*) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#5  0x00007fca56197ae8 in ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#6  0x00007fca562e8f40 in void tbb::interface6::internal::partition_type_base<tbb::interface6::internal::auto_partition_type>::execute<tbb::interface6::internal::start_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeElemAuxVarsThread, tbb::auto_partitioner const>, libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >(tbb::interface6::internal::start_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeElemAuxVarsThread, tbb::auto_partitioner const>&, libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>&) ()
   from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#7  0x00007fca562e90e2 in tbb::interface6::internal::start_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeElemAuxVarsThread, tbb::auto_partitioner const>::execute() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#8  0x00007fca570f515a in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all (this=0x7fca4b2d7e00, parent=..., child=0xa)
    at ../../src/tbb/custom_scheduler.h:438
#9  0x00007fca570f2d31 in tbb::internal::arena::process (this=0xb, s=...) at ../../src/tbb/arena.cpp:98
#10 0x00007fca570f0a75 in tbb::internal::market::process (this=0xb, j=...) at ../../src/tbb/market.cpp:465
#11 0x00007fca570ee254 in tbb::internal::rml::private_worker::run (this=0xb) at ../../src/tbb/private_server.cpp:274
#12 0x00007fca570ee186 in tbb::internal::rml::private_worker::thread_routine (arg=0xb) at ../../src/tbb/private_server.cpp:231
#13 0x00007fca4e385182 in start_thread (arg=0x7fca4aaf6700) at pthread_create.c:312
#14 0x00007fca4daa247d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

**Comment by permcody** [comment]
@friedmud - as you can see we have an analogous problem in `computeElemValue()` too.

**Comment by friedmud** [comment]
PR is in: https://github.com/libMesh/libmesh/pull/841

**Comment by friedmud** [comment]
@permcody Yeah: it's the same thing... it's calling the same function

**Comment by friedmud** [comment]
The reason we haven't tracked this down before is I just never saw a good stack trace.  Both of these are _perfect_.

**Comment by permcody** [comment]
Yeah, I'm thinking about adding a new capability to the TestHarness to grab
these "core" files when they appear and put them in the failed report. I'm
tired of these threaded failure. They've got to go!

BTW - I learned something on OS X that I probably should have know about
all this time but just figured out. That CrashReporter that runs when your
program crashes writes something similar to a core file to an area that is
read in the "Console" automatically. You can usually see good stack traces
on your Mac in Console under the "User Diagnostic Reports" section for your
program.

**Comment by brianmoose** [comment]
This just happened on the libmesh update PR: https://www.moosebuild.org/job/12295/

```
#0  0x00007f4a1a7e97f4 in libMesh::PetscVector<double>::operator()(unsigned int) const () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
(gdb) bt
#0  0x00007f4a1a7e97f4 in libMesh::PetscVector<double>::operator()(unsigned int) const () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#1  0x00007f4a1abff32a in MooseVariable::computeNodalValues() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#2  0x00007f4a1ab33da0 in SystemBase::reinitNodeFace(libMesh::Node const*, short, unsigned int) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#3  0x00007f4a1ac45b8d in FEProblem::reinitNodeFace(libMesh::Node const*, short, unsigned int) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#4  0x00007f4a1ac10b4f in ComputeNodalAuxBcsThread::onNode(__gnu_cxx::__normal_iterator<BndNode const* const*, std::vector<BndNode const*, std::allocator<BndNode const*> > >&) ()
   from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#5  0x00007f4a1ab73fb3 in ThreadedNodeLoop<libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>, __gnu_cxx::__normal_iterator<BndNode const* const*, std::vector<BndNode const*, std::allocator<BndNode const*> > > >::operator()(libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*> const&) ()
   from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#6  0x00007f4a1acffff6 in void tbb::interface6::internal::partition_type_base<tbb::interface6::internal::auto_partition_type>::execute<tbb::interface6::internal::start_reduce<libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>, ComputeNodalAuxBcsThread, tbb::auto_partitioner const>, libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*> >(tbb::interface6::internal::start_reduce<libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>, ComputeNodalAuxBcsThread, tbb::auto_partitioner const>&, libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>&) () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#7  0x00007f4a1ad00202 in tbb::interface6::internal::start_reduce<libMesh::StoredRange<MooseMesh::const_bnd_node_iterator, BndNode const*>, ComputeNodalAuxBcsThread, tbb::auto_partitioner const>::execute() () from /home/moosetest/civet/client_root_1/moose/framework/libmoose-opt.so.0
#8  0x00007f4a1b7bb15a in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all (this=0x7f4a0faf7e00, parent=..., child=0x0)
    at ../../src/tbb/custom_scheduler.h:438
#9  0x00007f4a1b7b8d31 in tbb::internal::arena::process (this=0xb, s=...) at ../../src/tbb/arena.cpp:98
#10 0x00007f4a1b7b6a75 in tbb::internal::market::process (this=0xb, j=...) at ../../src/tbb/market.cpp:465
#11 0x00007f4a1b7b4254 in tbb::internal::rml::private_worker::run (this=0xb) at ../../src/tbb/private_server.cpp:274
#12 0x00007f4a1b7b4186 in tbb::internal::rml::private_worker::thread_routine (arg=0xb) at ../../src/tbb/private_server.cpp:231
#13 0x00007f4a12b91e9a in start_thread (arg=0x7f4a0f313700) at pthread_create.c:308
#14 0x00007f4a122b938d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:112
#15 0x0000000000000000 in ?? ()
```

**Comment by permcody** [comment]
@friedmud - Looks like we didn't get it. Is it possible this thread race affects more than just the _get/_restore arrays you messed with? What about map_global_to_local_index:
https://github.com/libMesh/libmesh/blob/878e2dce447adbcd27f2b2eb308cdf89f3ad0a21/include/numerics/petsc_vector.h#L1132

**Comment by jwpeterson** [comment]
Hmm... I suppose there could be a race on the class variable `_array_is_present`?  Otherwise, `VecGetOwnershipRange()` seems pretty safe (?) and everything else is local variables.

**Comment by friedmud** [comment]
Craps

_array_is_present is guarded now.

Let me work on this a bit. I have time tonight. Can I abuse the testing
system by just putting in PRs with crazy libMesh branches?

**Comment by permcody** [comment]
Yes you can use the test systems all you want. Each time you push it
cancels the old job so you can't really clobber it.

**Comment by friedmud** [comment]
In trying to reproduce this I found something else.

@roystgnr this looks to be an issue with `constrain_element_matrix` in parallel.  I can reproduce it in `dbg` mode on linux with TBB in the `kernels/scalar_constraint` MOOSE test.  It only seems to happen when I go to many threads (8+).

```
Assertion `col_dofs.size() == matrix.n()' failed.
col_dofs.size() = 1
matrix.n() = 9


Stack frames: 26
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*)
2: libMesh::DofMap::constrain_element_matrix(libMesh::DenseMatrix<double>&, std::__debug::vector<unsigned int, std::allocator<unsigned int> >&, std::__debug::vector<unsigned int, std::allocator<unsigned int> >&, bool) const
3: Assembly::addJacobianBlock(libMesh::SparseMatrix<double>&, libMesh::DenseMatrix<double>&, std::__debug::vector<unsigned int, std::allocator<unsigned int> > const&, std::__debug::vector<unsigned int, std::allocator<unsigned int> > const&, double)
4: Assembly::addJacobianOffDiagScalar(libMesh::SparseMatrix<double>&, unsigned int)
5: FEProblem::addJacobianOffDiagScalar(libMesh::SparseMatrix<double>&, unsigned int, unsigned int)
6: NonlinearSystem::computeScalarKernelsJacobians(libMesh::SparseMatrix<double>&)
7: NonlinearSystem::computeJacobianInternal(libMesh::SparseMatrix<double>&)
8: NonlinearSystem::computeJacobian(libMesh::SparseMatrix<double>&)
9: FEProblem::computeJacobian(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::SparseMatrix<double>&)
10: Moose::compute_jacobian(libMesh::NumericVector<double> const&, libMesh::SparseMatrix<double>&, libMesh::NonlinearImplicitSystem&)
11: __libmesh_petsc_snes_jacobian
12: SNESComputeJacobian
13: SNESSolve_NEWTONLS
14: SNESSolve
15: libMesh::PetscNonlinearSolver<double>::solve(libMesh::SparseMatrix<double>&, libMesh::NumericVector<double>&, libMesh::NumericVector<double>&, double, unsigned int)
16: libMesh::NonlinearImplicitSystem::solve()
17: TimeIntegrator::solve()
18: NonlinearSystem::solve()
19: FEProblem::solve()
20: Steady::execute()
21: MooseApp::executeExecutioner()
22: MooseApp::run()
23: ../../../moose_test-dbg() [0x4139c0]
24: __libc_start_main
25: ../../../moose_test-dbg() [0x413609]
[0] ../src/base/dof_map_constraints.C, line 1734, compiled Mar  1 2016 at 19:34:54
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

**Comment by roystgnr** [comment]
That doesn't look like a constrain_element_matrix issue.  The preconditions of the function are being violated.  Is it possible that matrix or col_dofs is being shared between threads?

**Comment by friedmud** [comment]
@roystgnr Possible - I'll look into it after I figure out what's going on with this `operator()` business.  Thanks for weighing in.

**Comment by friedmud** [comment]
It's definite a race condition for `_array_is_present` here: https://github.com/libMesh/libmesh/blob/master/include/numerics/petsc_vector.h#L1139

Setting that to `if(false)` "fixes" it.  I'm still trying to work out exactly why.

It's a bugger to reproduce BTW.  Gotta run moose_test in a loop thousands of times before it happens...

**Comment by friedmud** [comment]
The only way I can see that `_array_is_present` could be wrong (and/or the cached `_first` and `_last` could be out of sync is if `_get_array()` and `_restore_array()` are called simultaneously (because of the doubly checked lock it is a possibility).

HOWEVER: that _should_ never happen.  In fact, I just added that would catch that and throw an error... and it DOES NOT happen.

So I'm stumped.

I can "fix it" by removing the use of cached `_first` and `_last` in `map_global_to_local_index()`... but I have this itching feeling that while it removes the segfault it doesn't actually fix the true problem.  BTW: There is no real reason to cache these values.  The `if()` test alone is probably more expensive than just calling `VecGetOwnershipRange()`  (that function is only a couple of lines long and just copies integers into the arguments that are passed to it).  But, like I say... I'm worried that may just be covering up the symptoms of a larger problem.

**Comment by friedmud** [comment]
I actually think this is a side-effect of memory operation ordering surrounding the doubly-checked lock.  There are technical reasons why a doubly-checked lock can have pathological race conditions.  Memory operations can play out in different orders for different threads as new values of shared memory gets propagated.  I've never run into this actually happening before though.

It looks like there is actually a fix for this in C++11: http://en.cppreference.com/w/cpp/atomic/memory_order

Here is a good writeup on the situation: http://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/

I can confirm that just locking all of `_get_array()` _does_ fix the issue... so I guess this really is the problem.  But now the question is: what to do about it?  Is _this_ the thing that finally pushes us to C++11?  Actually locking and unlocking for every `PetscVector::operator()` call would be completely fatal...

**Comment by permcody** [comment]
Well I've been reading about this issue and it sure is nasty. It's hard for me to believe that this is actually it but I'm willing to play along since we need a solution to this problem. 

I know how badly you'd like to move to C++11 but I'm not entirely convinced that it's required to fix this problem. There are indeed no portable ways to deal with this issue but there are practical ways of dealing with it and we are already at least part of the way down this path with the `atomic` abstraction in the `threads.h`. I can see that we have the `tbb/atomic` header pulled in. We also have the OS X specific atomic pulled in. It does appear that there aren't very many good solutions for this on Linux platforms though. I did run across a fix using thread local storage which is available in pthreads. However the solution may be staring us in the face. That Mintomic library looks pretty cool.

We could try C++11 specific code for the time being if you'd like in a branch and let it burn for several hours on one of our test systems to see if that's it, but it really shouldn't be too hard to finish up the abstracting we have going now.

The C++11 discussion should definitely continue. You've already managed to get the libMesh developers to reverse course once this week! :wink: 

I'd like @roystgnr to weigh in too.

**Comment by friedmud** [comment]
atomic is not enough (if you're thinking about making `is_array_present`
atomic)

Also, even if you make all of those variables (first, last, values, etc.)
atomic that would be slower than just locking and wouldn't fix the issue.

This really is the real problem. I traced it all out last night. Here's
what happens:

last = 0
first = 0
Is_array_present = 0

On all processors

Processor 0 gets into _get_array and grabs the lock and starts setting
values. It sets last, first and finally is_array_present...

The moment it sets is_array_present processor 1 hits the outer `if` of the
doubly checked lock in _get_array. It happens to get the updated value of
is_array_present but hasn't received the updated values of last and
first... so it skips _get_array and very quickly grabs values for last and
first that are still zero!

It is a very rare memory coherency issue. Processor 0 set values in this
order: first, last, is_array_present but they are being received on
Processor 1 in a different order.

For now, I vote that we just lock _get_array. It will slow things down, but
it will enforce correctness.

I can work up a solution that completely locks _get_array when you don't
have C++11 and uses the C++11 memory fences instead when they're available.
We could look at the differences in speed between the two approaches (and,
of course, check them for correctness).

I ran the completely locked _get_array case last night for about an hour
with one problem just looping (probably ran millions of times) without an
error. Without the lock it errors within about 3-5 minutes (still takes
thousands of runs).

**Comment by jwpeterson** [comment]
> For now, I vote that we just lock _get_array. It will slow things down, but it will enforce correctness.

I like this solution for the short term as well.  I'm curious just how bad of a performance hit we're talking about, but I'd be willing to accept quite a bit of slow down to avoid crashes!

**Comment by jwpeterson** [comment]
Also, knowing what we now know about the doubly-checked lock pattern, we should probably remove any others from MOOSE if they exist...

**Comment by permcody** [comment]
OK fair enough, I understand that the traditional atomic operation isn't enough but that's not due to a problem with atomic operations. It's due to the fact that in this particular pattern not all threads execute the atomic by design....

TBB's "atomic" however does handle the memory coherency issue: https://www.threadingbuildingblocks.org/docs/help/tbb_userguide/Design_Patterns/Lazy_Initialization.htm 

It looks like we have several things we need to do. We should do some timing and see what the penalty is for not using doubly-checked locks. If it shows up as a non-trivial expense the we'll know whether we need to get serious about a solution. C++11 clearly is one of those, but so is this Mintonic library.

**Comment by permcody** [comment]
Well :hankey: 
https://github.com/mintomic/mintomic/blob/master/README.md

**Comment by jwpeterson** [comment]
Also, I think we are also using the DCLP in libmesh when creating `RemoteElems`:
```
const Elem & RemoteElem::create ()
{
  if (remote_elem != libmesh_nullptr)
    return *remote_elem;

  RemoteElemMutex::scoped_lock lock(remote_elem_mtx);

  // check again - object could have been created while waiting
  // for the lock to acquire!
  if (remote_elem == libmesh_nullptr)
    remote_elem = new RemoteElem;

  return *remote_elem;
}
```

I know that helgrind complains about something to do with `Singleton`/`RemoteElem` creation (lock order violation) so I wonder if that's related...

```
==32216== ---Thread-Announcement------------------------------------------
==32216==
==32216== Thread #1 is the program's root thread
==32216==
==32216== ----------------------------------------------------------------
==32216==
==32216== Thread #1: lock order "0xABACD10 before 0xABAA364" violated
==32216==
==32216== Observed (incorrect) order is: acquisition of lock at 0xABAA364
==32216==    at 0x4C2ED8D: pthread_spin_lock (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==32216==    by 0x427295: libMesh::Threads::spin_mutex::lock() (threads.h:373)
==32216==    by 0x4272DB: libMesh::Threads::spin_mutex::scoped_lock::scoped_lock(libMesh::Threads::spin_mutex&) (threads.h:380)
==32216==    by 0x943D641: libMesh::Singleton::cleanup() (libmesh_singleton.C:98)
==32216==    by 0x9419E4B: libMesh::LibMeshInit::~LibMeshInit() (libmesh.C:629)
==32216==    by 0x6E44D9F: MooseInit::~MooseInit() (MooseInit.C:49)
==32216==    by 0x422468: main (main.C:39)
==32216==
==32216==  followed by a later acquisition of lock at 0xABACD10
==32216==    at 0x4C2ED8D: pthread_spin_lock (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==32216==    by 0x427295: libMesh::Threads::spin_mutex::lock() (threads.h:373)
==32216==    by 0x4272DB: libMesh::Threads::spin_mutex::scoped_lock::scoped_lock(libMesh::Threads::spin_mutex&) (threads.h:380)
==32216==    by 0x9789A3F: libMesh::RemoteElem::~RemoteElem() (remote_elem.C:62)
==32216==    by 0x9789AFB: libMesh::RemoteElem::~RemoteElem() (remote_elem.C:65)
==32216==    by 0x943D74E: libMesh::Singleton::cleanup() (libmesh_singleton.C:106)
==32216==    by 0x9419E4B: libMesh::LibMeshInit::~LibMeshInit() (libmesh.C:629)
==32216==    by 0x6E44D9F: MooseInit::~MooseInit() (MooseInit.C:49)
==32216==    by 0x422468: main (main.C:39)
==32216==
==32216== Required order was established by acquisition of lock at 0xABACD10
==32216==    at 0x4C2ED8D: pthread_spin_lock (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==32216==    by 0x427295: libMesh::Threads::spin_mutex::lock() (threads.h:373)
==32216==    by 0x4272DB: libMesh::Threads::spin_mutex::scoped_lock::scoped_lock(libMesh::Threads::spin_mutex&) (threads.h:380)
==32216==    by 0x9789B59: libMesh::RemoteElem::create() (remote_elem.C:74)
==32216==    by 0x97899C6: (anonymous namespace)::RemoteElemSetup::setup() (remote_elem.C:45)
==32216==    by 0x943D56D: libMesh::Singleton::setup() (libmesh_singleton.C:90)
==32216==    by 0x941895D: libMesh::LibMeshInit::LibMeshInit(int, char const* const*, ompi_communicator_t*) (libmesh.C:376)
==32216==    by 0x6E44D2C: MooseInit::MooseInit(int, char**, ompi_communicator_t*) (MooseInit.C:32)
==32216==    by 0x4223C0: main (main.C:25)
==32216==
==32216==  followed by a later acquisition of lock at 0xABAA364
==32216==    at 0x4C2ED8D: pthread_spin_lock (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==32216==    by 0x427295: libMesh::Threads::spin_mutex::lock() (threads.h:373)
==32216==    by 0x4272DB: libMesh::Threads::spin_mutex::scoped_lock::scoped_lock(libMesh::Threads::spin_mutex&) (threads.h:380)
==32216==    by 0x943D3C1: libMesh::Singleton::Singleton() (libmesh_singleton.C:66)
==32216==    by 0x9789E40: libMesh::RemoteElem::RemoteElem() (remote_elem.h:62)
==32216==    by 0x9789B91: libMesh::RemoteElem::create() (remote_elem.C:79)
==32216==    by 0x97899C6: (anonymous namespace)::RemoteElemSetup::setup() (remote_elem.C:45)
==32216==    by 0x943D56D: libMesh::Singleton::setup() (libmesh_singleton.C:90)
==32216==    by 0x941895D: libMesh::LibMeshInit::LibMeshInit(int, char const* const*, ompi_communicator_t*) (libmesh.C:376)
==32216==    by 0x6E44D2C: MooseInit::MooseInit(int, char**, ompi_communicator_t*) (MooseInit.C:32)
==32216==    by 0x4223C0: main (main.C:25)
==32216==
```

**Comment by friedmud** [comment]
@permcody The TBB example isn't what we have.  It's trying to initialize _one_ variable.  That can be done with the help of atomics.  The issue that we have is that we have one "guard" variable and we are initializing several other variables.  Even atomics won't fix this situation because the "guard" variable can be true on other processors before they've received the values of the other variables.  Even making them all atomics doesn't fix it (the atomic stores can still be received in a different order on the other processors).

The ONLY fix is to ensure that the memory operations happen in the same order on all processors which is what the memory fences do in C++11.  There is no simple, portable fix for this.

It is a memory instruction re-ordering issue between _multiple_ instructions... not just one.

@jwpeterson It's possible.

**Comment by friedmud** [comment]
@permcody: I take some of that back... it does sound like if we made _all_ of the variables TBB atomics then it would do the right thing.  Unfortunately, that would be slower than just locking the whole thing... and I'm not sure that it would work with our Pthread/OpenMP implementation of `atomic` either.

**Comment by permcody** [comment]
I'm coming to the same conclusion as well. We could use TBB to solve this
issue but that solution won't do us any good for the other two models.
Mintomic might work for the other models but is no longer supported.
Perhaps it could be used as a short term fix with a real plan towards
C++11, otherwise we _have_ to remove these double check locks!

**Comment by friedmud** [comment]
Yeah - and removing them sucks. I have an idea... let me try it out

**Comment by roystgnr** [comment]
One of the top hits about Doubly Checked Locking is a Dr. Dobbs article by Scott Meyers and Andrei Alexandrescu, which is entitled "C++ and the Perils of Double-Checked Locking" but which might as well be titled "You're Screwed Without Memory Barriers".

I honestly don't understand the "You can't safely do double-checked locking in portable pre-C++11!" grargh, though.  You can't spawn a thread in portable pre-C++11, so locking isn't a worry.  If you're using C++03-plus-something-that-allows-threads then odds are good that something-that-allows-threads also includes memory barriers.

E.g. You guys already found a memory barrier in TBB; and there are also a bunch of pthreads functions guaranteed to include memory fences:
http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_11

I'm not sure this route is going to be much less disappointing than "just lock the whole thing", though; apparently a single memory fence instruction can take hundreds of cycles on x86...

**Comment by permcody** [comment]
OK - well we need an action. I suggest we consider both: 
- Lock the whole thing for now
- Investigate abstracting and using memory fences later

**Comment by friedmud** [comment]
I have another idea... but I have a test tomorrow so I won't be able to try
it out until after that...

**Comment by permcody** [comment]
@friedmud We are holding off on the libMesh update. It's nearly 5:00 PM on Thursday here on a Friday off week. Let us know if you'll be able to get to this during the weekend. If not, we may merge libMesh on Monday to get the other threading fixes into MOOSE. I would really love to have this PETSc lock fixed though, even if we go for the lock-the-whole-thing approach. Have a good weekend!

**Comment by friedmud** [comment]
I will be able to fix it this weekend. It will be fixed one way or the
other...  Sorry for the delay!

BTW: My test went well ;-)

**Comment by friedmud** [comment]
Ignoring https://github.com/idaholab/moose/issues/6511 (fix that one later) and using https://github.com/idaholab/moose/pull/6510 and https://github.com/libMesh/libmesh/pull/853 I can now loop `run_tests --n-threads=6` indefinitely without a single error.

Once https://github.com/idaholab/moose/issues/6511 is taken care of... we may have licked this...
